### PR TITLE
Add `MacPorts` support to the build system

### PR DESCRIPTION
**Pull request by ghost**
# PR Summary

Add MacPorts support to the build system

## PR Context
MacPorts is one of the "Missing package managers for OS X".
The original build scripts only supported Homebrew and it is inadvised to have
both installed on a same system.
This change makes the build script use whatever package manager is currently
installed, or direct the user to download one of them if neither is present.
The build documentation has been updated to reflect that another package
manager is now also supported.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: build.psm1 [hunk 1]
```diff
@@ -1720,7 +1733,9 @@ function Start-PSBootstrap {
 
                 # Install dependencies
                 # ignore exitcode, because they may be already installed
-                Start-NativeExecution { brew install $Deps } -IgnoreExitcode
+                Start-NativeExecution {
+                    Invoke-Expression "$PackageManager install $Deps"
+                } -IgnoreExitcode
```
Comment (TravisEz13) [inline build.psm1]: Please remove the `Invoke-Expression`
```
                Start-NativeExecution {
                    &$PackageManager install $Deps
                } -IgnoreExitcode
```

Comment (TravisEz13) [reply build.psm1]: If that doesn't work
```
Start-NativeExecution ([ScriptBlock]::Create("$PackageManager install $Deps"))
```

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
See comments

File: build.psm1 (hunk 1 above)
Comment (Lucius-Q-User) [reply build.psm1]: Are you sure about that? The comment above explicitly says to use invoke-expression for sudo interpolation to work properly.

Comment (TravisEz13) [reply build.psm1]: New line 1756 we are doing what I'm suggesting.  I'm 100% sure one of these will work.

**Comment by TravisEz13** [comment]
There is a spelling error.

Here is my function to test spelling.

```powershell
function Test-Spelling
{
    param(
        [string[]]
        $Paths,
        [switch]
        $Fix

    )

    if(!(Get-Command mdspell -ErrorAction SilentlyContinue))
    {
        brew install yarn
        sudo yarn global add 'markdown-spellcheck@0.11.0'
    }

    $fileList = @()

    foreach($path in $Paths)
    {
        if($path -match '^\.[/\\]')
        {
            $fileList += ($path -replace '^\.[/\\]')
        }
        else {
            $fileList += $path
        }
    }

    $extraParams = @()

    if(!$Fix.IsPresent)
    {
        $extraParams += '--report'
    }

    Write-Verbose "Testing spelling for $fileList" -Verbose
    mdspell $fileList --ignore-numbers --ignore-acronyms @extraParams --en-us --no-suggestions
}
```

**Comment by Lucius-Q-User** [comment]
> There is a spelling error.
The script only complains about the words "2.x" (referring to the required .net core version) and "MacPorts". Looks like false positives to me.

**Comment by TravisEz13** [comment]
Spelling errors must be fixed for the PR to be accepted.  Please run the tool and add the words to the dictionary.

**Comment by TravisEz13** [comment]
There was a random failure verifying a URL.  I retried that task.

**Comment by TravisEz13** [comment]
@PoshChan Please remind me in 1 hour

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
Hold for compliance review

**Comment by TravisEz13** [comment]
@Lucius-Q-User To explain what triggered the review, MacPorts license is not clearly declared on GitHub (image below).  So, I'm waiting on a manual review of the use of this software in our project.  They don't give me a specific ETA.

# MacPro License
image

# MIT License
image

**Comment by PoshChan** [comment]
@TravisEz13, this is the reminder you requested 1 hour ago

File: build.psm1 (hunk 1 above)
Comment (vexx32) [reply build.psm1]: Pretty sure the `& $Packagemanager install @Deps` method would work just fine, if not better... all depends on the syntax expected by `brew` or `port` -- do they expect a space-delimited list of values, or a comma-delimited list of values for the package names?

Comment (Lucius-Q-User) [reply build.psm1]: Both of them want a space-delimited list of package names. But the problem with & is not the package list but rather the fact that port needs sudo to run, and if the user specifies -NoSudo, the $sudo expands to an empty string, which works fine with Invoke-Expression or ScriptBlock, but fails when called with &

Comment (vexx32) [reply build.psm1]: Oh, fair enough I suppose. You could just if/else that though. :)

Comment (TravisEz13) [reply build.psm1]: @vexx32 We have many other places we are creating script blocks... I'm fine with this for now.

**Comment by TravisEz13** [comment]
@Lucius-Q-User I'm monitoring the compliance review and I'd expect an answer within a week at this point.  Ping me if we don't get an answer by that time.

**Review by TravisEz13 — APPROVED** [review]

**Pull request merged**